File: interview/maxProduct.go
package interview

func maxProduct(nums []int) int {
	memo := make([][]int, 2)
	// memo[0][i] => i位置的最大值
	// memo[1][i] => i位置的最小值
	memo[0], memo[1] = make([]int, len(nums)), make([]int, len(nums))
	memo[0][0], memo[1][0] = nums[0], nums[0]
	max := nums[0]
	// condition 为true，返回n1,n2 大值
	If := func(condition bool, n1, n2 int) int {
		if (condition && n1 > n2) || (!condition && n1 < n2) {
			return n1
		}
		return n2
	}
	// dp
	for i := 1; i < len(nums); i++ {
		if nums[i] < 0 {
			memo[0][i] = If(true, memo[0][i-1]*nums[i], nums[i])
			memo[1][i] = If(false, memo[1][i-1]*nums[i], nums[i])
		} else {
			memo[0][i] = If(true, memo[1][i-1]*nums[i], nums[i])
			memo[1][i] = If(false, memo[0][i-1]*nums[i], nums[i])
		}
		max = If(true, max, memo[0][i])
	}
	return max
}

File: feishu/MaxProduct.java
/**
 * @author Zexho
 * @date 2022/3/30 11:31 PM
 */
public class MaxProduct {

    public int maxProduct(int[] nums) {
        int[][] arr = new int[2][nums.length];
        int max = arr[0][0] = arr[1][0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            arr[0][i] = Math.max(nums[i], nums[i] * arr[nums[i] < 0 ? 1 : 0][i - 1]);
            arr[1][i] = Math.min(nums[i], nums[i] * arr[nums[i] < 0 ? 0 : 1][i - 1]);
            max = Math.max(max, arr[0][i]);
        }
        return max;
    }

}
